File: python/loss.py
from layer import Layer
import numpy as np

class Loss(object):
    def __init__(self):
        pass

    def calc(self):
        pass

    def partial_derivative(self):
        pass

class MSE(Loss):
    def __init__(self):
        super(MSE, self).__init__()

    def calc(self, X, label, batch=1):
        if batch==1:
            if len(X.shape) != 1 or (len(label.shape) != 1 and label.shape[0] != 1):
                print "loss error"

        return np.sum(np.power(np.abs(X-label), 2))*0.5

    def partial_derivative(self, X, label):
        return - (label - X)


class CrossEntropy(Loss):
    def __init__(self):
        super(CrossEntropy, self).__init__()

    def calc(self, X, label, batch=1):
        print X
        return -np.sum(label * np.log(X+10e-20) + (1-label)*np.log(1-X), axis=0)

    def partial_derivative(self, X, label):
        return (X-label)/(X * (1 - X))

File: cpp/loss.cc
#include "loss.h"
#include <cmath>
#include <stdlib.h>

Loss::Loss() {}
Loss::~Loss() {
    this->D.clear();
    this->D.shrink_to_fit();
}

int Loss::configure(int batch, Layer* prevLayer) {
    if (prevLayer != nullptr) {
        this->prevLayer = prevLayer;
    }
    this->batch = batch;
    this->batch_inv = 1/batch;
    this->D.resize(this->batch*this->prevLayer->units);
    return 1;
}

MSE::MSE() {}
MSE::~MSE() {}

int MSE::configure(int batch, Layer* prevLayer) {
    return Loss::configure(batch, prevLayer);
}

float MSE::error(vector<float> *x, int* label) {
    float e = 0;
    #pragma omp  parallel for reduction(+:e)
    for (int i = 0; i < this->batch*this->prevLayer->units; i++) {
        e += std::pow(x->at(i)-(float)label[i], 2)*0.5*this->batch_inv;
    }
    return e;
}

void MSE::partial_derivative(vector<float>* x, int* label) {
    #pragma omp  parallel for
    for (int i = 0; i < this->batch*this->prevLayer->units; i++) {
        this->D[i] = - ((float)label[i] - x->at(i));
    }
    return;
}
